New blog on self-managed OpenSearch Dashboards (#2852)
opensearch-project/project-website: _posts/2024-04-30-A-comprehensive-guide-to-setup-and-connect-self-managed-dashboards-with-Amazon-OpenSearch-domain.md

---
layout: post
title:  "A comprehensive guide to setting up and connecting self-managed OpenSearch Dashboards with an Amazon OpenSearch Service domain"
authors:
 - mkbn

> date: 2024-04-30

Reviewer: Please updated blog date to 2024-05-15

categories:
 - technical-post

> meta_keywords: OpenSearch dashboards, Amazon OpenSearch service, self-managed, downtime issues, Mitigation

Reviewer: Please update the meta with the following: 

**meta_keywords:** OpenSearch Dashboards, Amazon OpenSearch Service, self-managed dashboards, blue/green deployments

**meta_description**: Read this guide to blue/green deployments to learn how you can set up self-managed dashboards in Amazon OpenSearch, including no authentication, basic authentication, and SAML authentication.

To maintain continuous access to dashboards and visualizations during blue/green deployment, users can implement a workaround by setting up and connecting a self-managed Dashboards instance with a managed service domain. By using self-managed Dashboards instances, users can ensure continuous access to their dashboards and visualizations throughout the blue/green deployment process, minimizing downtime and mitigating any potential impact to business operations.

This solution currently supports three different methods of authentication:
* No authentication
* HTTP basic authentication
* SAML authentication

> [!IMPORTANT]
> It is vital to choose the same major version of self-managed OpenSearch Dashboards as the source managed service domain across all supported methods (for example, while upgrading from 1.3 to 2.11, self-managed Dashboards should be on version 1.3). For Docker images, see the [Docker images repository](https://hub.docker.com/r/opensearchproject/opensearch-dashboards/tags).

# Setting up self-managed Dashboards in an Amazon EC2–hosted Docker container: No authentication

## Prerequisite

> AWS-managed OpenSearch domain without any authentication method enabled and is accompanied by the following domain access policy:

Reviewer 2: ```suggestion
An AWS-managed OpenSearch domain without any authentication method enabled, accompanied by the following domain access policy, is required:
```

```
FROM opensearchproject/opensearch-dashboards:2.5.0
RUN /usr/share/opensearch-dashboards/bin/opensearch-dashboards-plugin remove securityDashboards
COPY --chown=opensearch-dashboards:opensearch-dashboards opensearch_dashboards.yml /usr/share/opensearch-dashboards/config/
```
3. Run this command `docker build --tag=opensearch-dashboards-no-security .` to build a new Docker image with security plugin removed.
4. Validate whether the new image has been created by running the `docker images` command.
5. In the following sample `docker-compose.yml` file, change the Dashboards image name from `opensearchproject/opensearch-dashboards:2.5.0` to `opensearch-dashboards-no-security` and remove the username and password fields:
```yml
version: '3'
services:
  opensearch-dashboards:
    image: opensearchproject/opensearch-dashboards:2.5.0
    container_name: opensearch-dashboards
    ports:
      - 5601:5601
    expose:
      - "5601"
    environment:
      OPENSEARCH_HOSTS: '["https://success-2-ce6hkjt5gh.ap-south-1.es.amazonaws.com"]'
      OPENSEARCH_USERNAME: 'xxx'
      OPENSEARCH_PASSWORD: 'xxxx'
    networks:
      - opensearch-net
networks:
  opensearch-net:
```

> 6. The new `docker-compose-no-security.yml` file has now been created and should appear similar to the following file. Now run the `docker-compose up` command to run the containers with the new image. Then you can access the self-managed Dashboards instances by hitting the Amazon Elastic Compute Cloud (Amazon EC2) endpoint with port `5601`. By doing so, you can conveniently view and interact with all the saved objects.

Reviewer 2: ```suggestion
6. The new `docker-compose-no-security.yml` file has now been created and should appear similar to the following file. Now run the `docker-compose up` command to run the containers with the new image. Then you can access the self-managed Dashboards instances by connecting to the Amazon Elastic Compute Cloud (Amazon EC2) endpoint with port `5601`. By doing so, you can conveniently view and interact with all the saved objects.
```

```yml
version: '3'
services:
  opensearch-dashboards:
    image: opensearch-dashboards-no-security
    container_name: opensearch-dashboards
    ports:
      - 5601:5601
    expose:
      - "5601"
    environment:
      OPENSEARCH_HOSTS: '["https://success-2-ce6hkjt5gh.ap-south-1.es.amazonaws.com"]'
    networks:
      - opensearch-net
networks:
  opensearch-net:
```
# 
# Setting up self-managed Dashboards instances in Amazon ECS and in an EC2-hosted Docker container: HTTP basic authentication

## Prerequisite
The AWS-managed OpenSearch domain must incorporate fine-grained access control (FGAC) with HTTP basic authentication, ensuring that a primary user is created in the internal user database. For more information, see [this tutorial](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/fgac-http-auth.html).

## Spinning up a self-managed Dashboards instance in Amazon ECS
1. Create a task within Amazon Elastic Container Service (Amazon ECS) on AWS Fargate using the Dashboards Docker image.

> 2. When creating a task, under "container definition" in "port mapping", make sure the container ports `5601` and `9200` are added.

Reviewer 2: ```suggestion
2. When creating a task, under `container definition` in `port mapping`, make sure the container ports `5601` and `9200` are added.
```

> 2. When creating a task, under "container definition" in "port mapping", make sure the container ports `5601` and `9200` are added.

Reviewer 2: Re: your response to my comment on this, we actually do use code font for field names 😄.

Author: oh ok

> 3. Under environment variables, add the mandatory keys and values specified in [this document](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/dashboards.html#dashboards-local) to seamlessly connect with the managed service domain.

Reviewer 2: ```suggestion
3. Under `environment variables`, add the mandatory keys and values specified in [this document](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/dashboards.html#dashboards-local) to seamlessly connect with the managed service domain.
```

...
            "portMappings": [
                {
                    "name": "dash-5601-tcp",
                    "containerPort": 5601,
                    "hostPort": 5601,
                    "protocol": "tcp",
                    "appProtocol": "http"
                },
                {
                    "name": "dash-9200-tcp",
                    "containerPort": 9200,
                    "hostPort": 9200,
                    "protocol": "tcp",
                    "appProtocol": "http"
                }
            ],
            "essential": true,
            "environment": [
                {
                    "name": "OPENSEARCH_USERNAME",
                    "value": "xxx"
                },
                {
                    "name": "OPENSEARCH_HOSTS",
                    "value": "https://success-2-ce6hkjt5gh.ap-south-1.es.amazonaws.com"
                },
                {
                    "name": "OPENSEARCH_PASSWORD",
                    "value": "xxxx"
                }
            ],
...
}
```

4. Create a service using the previously created task within the same virtual private cloud (VPC) and subnet where the OpenSearch Service domain is operating.

> 5. Access the self-managed Dashboards instances by hitting the public endpoint of the running task in Amazon ECS on AWS Fargate. By doing so, you can conveniently view and interact with all the saved objects in accordance with the FGAC settings.

Reviewer 2: ```suggestion
5. Access the self-managed Dashboards instances by connecting to the public endpoint of the running task in Amazon ECS on AWS Fargate. By doing so, you can conveniently view and interact with all the saved objects in accordance with the FGAC settings.
```

      OPENSEARCH_USERNAME: 'xxx'
      OPENSEARCH_PASSWORD: 'xxxx'
    networks:
      - opensearch-net
networks:
  opensearch-net:
```
4. To enable TLS, add the attributes specified [here](https://opensearch.org/docs/latest/install-and-configure/install-dashboards/tls/) as environment variables.

> [!TIP]
> One notable advantage of setting up a self-managed Dashboards instance is that when it is deployed on `AWS ECS Fargate`, it generates a public IP address. This allows the self-managed Dashboards instance to be accessed over the internet without the need to set up a reverse proxy. As a result, the OpenSearch domains will be within the VPC, and the self-managed Dashboards instances will be publicly available, enabling seamless connectivity and eliminating the complexity of configuring additional infrastructure components. This simplifies the setup process and provides convenient access to the Dashboards instances from anywhere on the internet without compromising security or requiring additional network configurations.

# 
# Setting up a self-managed Dashboards instance in an EC2-hosted container: SAML authentication

## Prerequisite

> AWS managed OpenSearch domain with SAML authentication enabled. [Reference here](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/saml.html

Reviewer 2: ```suggestion
An AWS-managed OpenSearch domain with SAML authentication enabled is required. For more information, see [SAML authentication for OpenSearch Dashboards](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/saml.html).
```

)

## Spinning up a self-managed Dashboards instance in an EC2-hosted container
1. Create an EC2 instance within the same VPC where the managed OpenSearch Service domain is operating to configure the self-managed Dashboards instance and capture its endpoint.
2. Create a new application in your `IDP` with the self-managed Dashboards endpoint, which generates new identity provider (IdP) metadata.

> 3. Copy the IdP metadata of the newly created application and paste it into the IdP metadata text box found in the `Configure identity provider (IdP)` section on the security configuration tab of the managed service domain in the AWS Management Console. The following is the sample IdP metadata XML:

Reviewer 2: ```suggestion
3. Copy the IdP metadata of the newly created application and paste it into the IdP metadata text box found in the `Configure identity provider (IdP)` section on the **Security Configuration** tab of the managed service domain in the AWS Management Console. The following is the sample IdP metadata XML:
```

```yml
opensearch.hosts: [https://localhost:9200]
opensearch.ssl.verificationMode: none
opensearch.username: kibanaserver
opensearch.password: kibanaserver
opensearch.requestHeadersWhitelist: [authorization, securitytenant]
opensearch_security.multitenancy.enabled: true
opensearch_security.multitenancy.tenants.preferred: [Private, Global]
opensearch_security.readonly_mode.roles: [kibana_read_only]
# Use this setting if you are running opensearch-dashboards without https
opensearch_security.cookie.secure: false
server.host: '0.0.0.0'
opensearch_security.auth.type: "saml"
server.xsrf.whitelist: ["/_opendistro/_security/saml/acs", "/_opendistro/_security/saml/logout"]
```

> 8. After restarting the container, you can access the self-managed Dashboards instance by hitting the EC2 endpoint with port `5601`. By doing so, you can view and interact with all the saved objects in accordance with the FGAC settings and SAML authentication.

Reviewer 2: ```suggestion
8. After restarting the container, you can access the self-managed Dashboards instance by connecting to the EC2 endpoint with port `5601`. By doing so, you can view and interact with all the saved objects in accordance with the FGAC settings and SAML authentication.
```

To maintain continuous access to the dashboards and visualizations during blue/green deployment, customers can implement a workaround by setting up and connecting a self-managed OpenSearch Dashboards with the managed service domain. By utilizing self-managed instances of the OpenSearch Dashboards, customers can ensure continuous access to their dashboards and visualizations throughout the blue/green deployment process, minimizing downtime and mitigating any potential impact on the business operations.

This solution currently supports three different methods of authentication
* No authentication
* HTTP basic authentication
* SAML authentication

> [!IMPORTANT]

> > Choosing the identical major version of self-managed dashboards as the source managed service domain you plan to link them with is vital across all supported methods (e.g. while upgrading from 1.3 to 2.11, self-managed dashboards should be in 1.3) - [Docker images repo link](https://hub.docker.com/r/opensearchproject/opensearch-dashboards/tags)

Author: Nice, This is clear and retain the meaning too...perfect!

# Guide to setup self-managed dashboards in EC2 hosted docker container - No authentication

## Prerequisite

> AWS managed OpenSearch domain without any authentication method enabled and is accompanied by the following domain access policy.

Author: Yes, thats right. 
AWS-managed OpenSearch domain without any authentication method enabled, accompanied by the following domain access policy, is required

```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": "es:*",
      "Resource": "arn:aws:es:ap-south-1:765423874566:domain/no-security/*"
    }
  ]
}
```
> [!CAUTION]
> To establish a connection with the managed OpenSearch domain, it's necessary to uninstall the security plugin from self-managed dashboards. Otherwise, the Dashboards' security plugin will anticipate a secured domain and will fail to make a connection

## Steps to remove the security plugin and spin up a self-managed dashboards
1. Remove all Security plugin configurations from opensearch_dashboards.yml or place the below example file in the same folder as the Dockerfile
```yml
server.name: opensearch-dashboards
server.host: "0.0.0.0"
opensearch.hosts: http://localhost:9200
```
2. Create a new Dockerfile like below
```
FROM opensearchproject/opensearch-dashboards:2.5.0
RUN /usr/share/opensearch-dashboards/bin/opensearch-dashboards-plugin remove securityDashboards
COPY --chown=opensearch-dashboards:opensearch-dashboards opensearch_dashboards.yml /usr/share/opensearch-dashboards/config/
```

> 3. Run this command `docker build --tag=opensearch-dashboards-no-security .` to build a new Docker image with security plugin removed.

Author: `docker build --tag=opensearch-dashboards-no-security .` - This is the command, there should be a period at the end of the command

> 3. Run this command `docker build --tag=opensearch-dashboards-no-security .` to build a new Docker image with security plugin removed.

Reviewer 2: Got ya. Please commit the other two changes.

> 3. Run this command `docker build --tag=opensearch-dashboards-no-security .` to build a new Docker image with security plugin removed.

Reviewer 2: This one still needs to be committed.

4. Validate if the new image is created by running the `docker images` command
5. In the below sample `docker-compose.yml`, change the dashboards' image name from `opensearchproject/opensearch-dashboards:2.5.0` to `opensearch-dashboards-no-security` and remove the username and password fields.
```yml
version: '3'
services:
  opensearch-dashboards:
    image: opensearchproject/opensearch-dashboards:2.5.0
    container_name: opensearch-dashboards
    ports:
      - 5601:5601
    expose:
      - "5601"
    environment:
      OPENSEARCH_HOSTS: '["https://success-2-ce6hkjt5gh.ap-south-1.es.amazonaws.com"]'
      OPENSEARCH_USERNAME: 'xxx'
      OPENSEARCH_PASSWORD: 'xxxx'
    networks:
      - opensearch-net
networks:
  opensearch-net:
```

> 6. The new `docker-compose-no-security.yml` file is now created which should look like the below. Now run the `docker-compose up` command to run the containers with new image. Now customer can access the self-managed dashboards by hitting the EC2 endpoint with port `5601`. By doing so, you can conveniently view and interact with all the saved objects

Author: Can we say "connecting to the EC2 endpoint"

> 6. The new `docker-compose-no-security.yml` file is now created which should look like the below. Now run the `docker-compose up` command to run the containers with new image. Now customer can access the self-managed dashboards by hitting the EC2 endpoint with port `5601`. By doing so, you can conveniently view and interact with all the saved objects

Reviewer 2: Yes

```yml
version: '3'
services:
  opensearch-dashboards:
    image: opensearch-dashboards-no-security
    container_name: opensearch-dashboards
    ports:
      - 5601:5601
    expose:
      - "5601"
    environment:
      OPENSEARCH_HOSTS: '["https://success-2-ce6hkjt5gh.ap-south-1.es.amazonaws.com"]'
    networks:
      - opensearch-net
networks:
  opensearch-net:
```
# 
# Guide to setup self-managed dashboards in ECS & in EC2 hosted docker container - HTTP basic authentication

## Prerequisite

> AWS managed OpenSearch domain must incorporate Fine-Grained Access Control (FGAC) with HTTP basic authentication, ensuring that a master user is created in the internal user database.

Reviewer 2: ```suggestion
The AWS-managed OpenSearch domain must incorporate fine-grained access control (FGAC) with HTTP basic authentication, ensuring that a primary user is created in the internal user database. For more information, see [this tutorial](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/fgac-http-auth.html).
```

> [Reference here](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/fgac-http-auth.html)

Reviewer 2: ```suggestion
```

> ## Steps to spin up a self-managed dashboards in ECS

Reviewer 2: ```suggestion
## Spinning up a self-managed Dashboards instance in Amazon ECS
```

> 1. Create a task within ECS Fargate using the dashboards docker image.

Reviewer 2: ```suggestion
1. Create a task within Amazon Elastic Container Service (Amazon ECS) on AWS Fargate using the Dashboards Docker image.
```

> 2. When creating a task, under "container definition" in "port mapping" make sure the container ports `5601` and `9200` are added.

Reviewer 2: ```suggestion
2. When creating a task, under "container definition" in "port mapping", make sure the container ports `5601` and `9200` are added.
```

> 2. When creating a task, under "container definition" in "port mapping" make sure the container ports `5601` and `9200` are added.

Reviewer 2: Rather than using quotes, should "container definition" and "port mapping" be either bold or in code font?

Author: It is a field name so I think we can Bold those but not in code format.

> 3. Under environment variables, add the mandatory keys and values mentioned in this [doc](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/dashboards.html#dashboards-local) to seamlessly connect with managed service domain.

Reviewer 2: ```suggestion
3. Under environment variables, add the mandatory keys and values specified in [this document](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/dashboards.html#dashboards-local) to seamlessly connect with the managed service domain.
```

> 3. Under environment variables, add the mandatory keys and values mentioned in this [doc](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/dashboards.html#dashboards-local) to seamlessly connect with managed service domain.

Reviewer 2: Should "environment variables" be either bold or in code font?

Author: Lets do BOLD

> [!NOTE]

> > Below is the sample `task.json` file from the ECS task defenition to help understand the environment variables that has to be set while creating tasks

Reviewer 2: ```suggestion
> The following sample `task.json` file from the Amazon ECS task definition shows the environment variables that have to be set while creating tasks:
```

```json
{
...
            "portMappings": [
                {
                    "name": "dash-5601-tcp",
                    "containerPort": 5601,
                    "hostPort": 5601,
                    "protocol": "tcp",
                    "appProtocol": "http"
                },
                {
                    "name": "dash-9200-tcp",
                    "containerPort": 9200,
                    "hostPort": 9200,
                    "protocol": "tcp",
                    "appProtocol": "http"
                }
            ],
            "essential": true,
            "environment": [
                {
                    "name": "OPENSEARCH_USERNAME",
                    "value": "xxx"
                },
                {
                    "name": "OPENSEARCH_HOSTS",
                    "value": "https://success-2-ce6hkjt5gh.ap-south-1.es.amazonaws.com"
                },
                {
                    "name": "OPENSEARCH_PASSWORD",
                    "value": "xxxx"
                }
            ],
...
}
```

> 4. Create a service using the previously created task within the identical VPC and subnet where the managed service OpenSearch domain is operating.

Reviewer 2: ```suggestion
4. Create a service using the previously created task within the same virtual private cloud (VPC) and subnet where the OpenSearch Service domain is operating.
```

> 5. Access the self-managed dashboards by hitting the public endpoint of the running task in ECS Fargate. By doing so, you can conveniently view and interact with all the saved objects in accordance with the Fine-Grained Access Control settings.

Reviewer 2: ```suggestion
5. Access the self-managed Dashboards instances by hitting the public endpoint of the running task in Amazon ECS on AWS Fargate. By doing so, you can conveniently view and interact with all the saved objects in accordance with the FGAC settings.
```

> 5. Access the self-managed dashboards by hitting the public endpoint of the running task in ECS Fargate. By doing so, you can conveniently view and interact with all the saved objects in accordance with the Fine-Grained Access Control settings.

Reviewer 2: Please use a different/more precise word than "hitting".

Author: connecting to the EC2 endpoint

> ## Steps to spin up a self-managed dashboards in EC2 hosted container

Reviewer 2: ```suggestion
## Spinning up a self-managed Dashboards instance in an EC2-hosted container
```

> 1. Deploy an EC2 instance in the same VPC and subnet as the managed OpenSearch service domain.

Reviewer 2: ```suggestion
1. Deploy an EC2 instance in the same VPC and subnet as the managed OpenSearch Service domain.
```

> 2. Set up Docker/Kubernetes and its dependencies on the instance.

Reviewer 2: ```suggestion
2. Set up Docker/Kubernetes and their dependencies on the instance.
```

> 3. Utilize the below `docker-compose.yml` file to launch a self-managed dashboards container. After the container is running, you can easily access and interact with all the saved objects.

Reviewer 2: ```suggestion
3. Use the following `docker-compose.yml` file to launch a self-managed Dashboards container. After the container is running, you can easily access and interact with all the saved objects.
```

```yml
version: '3'
services:
  opensearch-dashboards:
    image: opensearchproject/opensearch-dashboards:2.5.0
    container_name: opensearch-dashboards
    ports:
      - 5601:5601
    expose:
      - "5601"
    environment:
      OPENSEARCH_HOSTS: '["https://success-2-ce6hkjt5gh.ap-south-1.es.amazonaws.com"]'
      OPENSEARCH_USERNAME: 'xxx'
      OPENSEARCH_PASSWORD: 'xxxx'
    networks:
      - opensearch-net
networks:
  opensearch-net:
```

> 4. To enable TLS, add the attributes mentioned in this [link](https://opensearch.org/docs/latest/install-and-configure/install-dashboards/tls/) as environment variables.

Reviewer 2: ```suggestion
4. To enable TLS, add the attributes specified [here](https://opensearch.org/docs/latest/install-and-configure/install-dashboards/tls/) as environment variables.
```

> [!TIP]

> > One notable advantage of setting up a self-managed dashboards is that when it is deployed on `AWS ECS Fargate`, it generates a public IP. This allows the self-managed dashboards to be accessed over the internet without the need for setting up a reverse proxy. As a result, the OpenSearch domains will be within the VPC and self-managed dashboards will be available in public, enabling seamless connectivity and eliminating the complexity of configuring additional infrastructure components. This simplifies the setup process and provides convenient access to the dashboards from anywhere on the internet without compromising security or requiring additional network configurations.

Reviewer 2: ```suggestion
> One notable advantage of setting up a self-managed Dashboards instance is that when it is deployed on `AWS ECS Fargate`, it generates a public IP address. This allows the self-managed Dashboards instance to be accessed over the internet without the need to set up a reverse proxy. As a result, the OpenSearch domains will be within the VPC, and the self-managed Dashboards instances will be publicly available, enabling seamless connectivity and eliminating the complexity of configuring additional infrastructure components. This simplifies the setup process and provides convenient access to the Dashboards instances from anywhere on the internet without compromising security or requiring additional network configurations.
```

> > One notable advantage of setting up a self-managed dashboards is that when it is deployed on `AWS ECS Fargate`, it generates a public IP. This allows the self-managed dashboards to be accessed over the internet without the need for setting up a reverse proxy. As a result, the OpenSearch domains will be within the VPC and self-managed dashboards will be available in public, enabling seamless connectivity and eliminating the complexity of configuring additional infrastructure components. This simplifies the setup process and provides convenient access to the dashboards from anywhere on the internet without compromising security or requiring additional network configurations.

Reviewer 2: Is `AWS ECS Fargate` intentionally in code font, or should it be "Amazon ECS on AWS Fargate" (https://docs.aws.amazon.com/AmazonECS/latest/developerguide/AWS_Fargate.html)?

Author: This is just a generic term and if "Amazon ECS on AWS Fargate" is the right set of words recommended, lets stick to this.

# 

> # Guide to setup self-managed dashboards in EC2 hosted container - SAML authentication

Reviewer 2: ```suggestion
# Setting up a self-managed Dashboards instance in an EC2-hosted container: SAML authentication
```

## Prerequisite

> AWS managed OpenSearch domain with SAML authentication enabled. [Reference here](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/saml.html

Reviewer 2: This should be written as a complete sentence. Do we mean something like "An AWS-managed OpenSearch domain with SAML authentication enabled is required"?

Author: Yes

)

> ## Steps to spin up a self-managed dashboards in EC2 hosted container

Reviewer 2: ```suggestion
## Spinning up a self-managed Dashboards instance in an EC2-hosted container
```

> 1. Create an EC2 instance within the identical VPC where the managed service OpenSearch domain is operating to configure the self-managed dashboards and capture its endpoint.

Reviewer 2: ```suggestion
1. Create an EC2 instance within the same VPC where the managed OpenSearch Service domain is operating to configure the self-managed Dashboards instance and capture its endpoint.
```

> 2. Create a new Application in your `IDP` with the self-managed dashboards endpoint which would generate a new IDP metadata

Reviewer 2: ```suggestion
2. Create a new application in your `IDP` with the self-managed Dashboards endpoint, which generates new identity provider (IdP) metadata.
```

> 3. Copy the IDP metadata of the newly created application and paste it into the IDP metadata text box found in the `Configure identity provider (IdP)` section within the security configuration tab of the managed service domain in aws console. Below is the sample IDP metadata xml.

Reviewer 2: ```suggestion
3. Copy the IdP metadata of the newly created application and paste it into the IdP metadata text box found in the `Configure identity provider (IdP)` section on the security configuration tab of the managed service domain in the AWS Management Console. The following is the sample IdP metadata XML:
```

> 3. Copy the IDP metadata of the newly created application and paste it into the IDP metadata text box found in the `Configure identity provider (IdP)` section within the security configuration tab of the managed service domain in aws console. Below is the sample IDP metadata xml.

Author: Sure

o5WxDfwfq8zds9dYV0YGTzUVQruLF9VpPIS/6QCOmVmbw6IP8nIIQZjwHBwCHCK3/ArLcYPIL28S
qA0i/ueP3VQWZcljAL3WRW0hUrupOW+sK5nIdF0Ac1OZYQIDAQABMA0GCSqGSIb3DQEBCwUAA4IB
AQBOJM5K86/mJx0zlM1dYmP/PbyUpA+QDSi7aNaYJ06tGomIWHyA8wyw0+dMy7S2ZzSm/buXUv/w
Bgn+nueNrZY5+cOLLW8DSayGG0lZanTgtiCqA7JuKgzwxXmpsld1d7JgQ+EshCNLvF8c3iR47/+R
/rTp7aZ/jn3c+BBynqTQX/2aYWVizQyAPeZjWPZbTjy1kunUTdv6rhLEP+HizH8HN7tCPf1l4HZS
OzAwZlwvGWNaT3kaLtjdLmFjlDV5PUMiQdBf6DKihH8fdQjty/vbswxqfMGj0aSppxzXn0XG1kwH
IK5Y04uMGfRjcE+cPA/vPCKPxh/sgB0n6GaJCIDI</ds:X509Certificate></ds:X509Data></ds:KeyInfo></md:KeyDescriptor><md:NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified</md:NameIDFormat><md:NameIDFormat>urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress</md:NameIDFormat><md:SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="https://trial-8803933.okta.com/app/trial-8803933_2325vpc_1/exk5o8zomj6eLo2an697/sso/saml"/><md:SingleSignOnService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Redirect" Location="https://trial-8803933.okta.com/app/trial-8803933_2325vpc_1/exk5o8zomj6eLo2an697/sso/saml"/></md:IDPSSODescriptor></md:EntityDescriptor>
```

> 4. Replace the self-managed dashboards url in the security configuration file with the self-managed dashboards endpoint. The purpose of this change is to guarantee that after the user is authenticated by IDP, the redirection occurs to the self-managed dashboards instead of the managed service dashboards.

Reviewer 2: ```suggestion
4. Replace the self-managed Dashboards URL in the security configuration file with the self-managed Dashboards endpoint. The purpose of this change is to guarantee that after the user is authenticated by IdP, they are redirected to the self-managed Dashboards instance instead of the managed Dashboards instance.
```

> [!IMPORTANT]

> > Customers do not have access to modify the security configuration file hence raise a support case to AWS support to request a change to the self-managed dashboards URL endpoint. After the AWS support completes your request, you can check the new endpoint by running API call ```_opendistro/_security/api/securityconfig```, customer can validate the `kibana_url` changes in security configuration file.

Reviewer 2: ```suggestion
> Users do not have access to modify the security configuration file, so you will need to raise a support case with AWS Support to request a change to the self-managed Dashboards URL endpoint. After AWS Support completes your request, you can check the new endpoint by running the API call ```_opendistro/_security/api/securityconfig``` and validate the `kibana_url` changes in the security configuration file.
```

> 5. Install docker and its dependencies on the EC2 instance

Reviewer 2: ```suggestion
5. Install Docker and its dependencies on the EC2 instance.
```

> 6. Use the below `docker-compose.yml` file and run the self-managed dashboards

Reviewer 2: ```suggestion
6. Use the following `docker-compose.yml` file and run the self-managed Dashboards instance:
```

```yml
version: '3'
services:
  opensearch-dashboards:
    image: opensearchproject/opensearch-dashboards:2.9.0
    container_name: opensearch-dashboards
    ports:
      - 5601:5601
    expose:
      - "5601"
    environment:
      OPENSEARCH_HOSTS: '["https://success-2-ce6hkjt5gh.ap-south-1.es.amazonaws.com"]'
      OPENSEARCH_USERNAME: 'xxx'
      OPENSEARCH_PASSWORD: 'xxxx'
    networks:
      - opensearch-net
networks:
  opensearch-net:
```

> 7. After the container is up and running, access it by using the command `docker exec -it <CONTAINER ID> bash` and then modify the `opensearch_dashboards.yml` file by adding the SAML specific attributes. Once the modifications are made, restart the container using `docker restart <CONTAINER ID>`. Find below the sample `opensearch_dashboards.yml` for reference. [Additional Reference here](https://opensearch.org/docs/latest/security/authentication-backends/saml/#opensearch-dashboards-configuration)

Reviewer 2: ```suggestion
7. After the container is up and running, access it by using the command `docker exec -it <CONTAINER ID> bash` and then modify the `opensearch_dashboards.yml` file by adding the SAML-specific attributes. Once the modifications are made, restart the container using `docker restart <CONTAINER ID>`. See the following sample `opensearch_dashboards.yml` file as a reference. See [OpenSearch Dashboards configuration](https://opensearch.org/docs/latest/security/authentication-backends/saml/#opensearch-dashboards-configuration) for more information.
```

```yml
opensearch.hosts: [https://localhost:9200]
opensearch.ssl.verificationMode: none
opensearch.username: kibanaserver
opensearch.password: kibanaserver
opensearch.requestHeadersWhitelist: [authorization, securitytenant]
opensearch_security.multitenancy.enabled: true
opensearch_security.multitenancy.tenants.preferred: [Private, Global]
opensearch_security.readonly_mode.roles: [kibana_read_only]
# Use this setting if you are running opensearch-dashboards without https
opensearch_security.cookie.secure: false
server.host: '0.0.0.0'
opensearch_security.auth.type: "saml"
server.xsrf.whitelist: ["/_opendistro/_security/saml/acs", "/_opendistro/_security/saml/logout"]
```

> 8. Post container restart you can access the self-managed dashboards by hitting the EC2 endpoint with port 5601. By doing so, you can view and interact with all the saved objects in accordance with the Fine-Grained Access Control settings and SAML authentication.

Reviewer 2: ```suggestion
8. After restarting the container, you can access the self-managed Dashboards instance by hitting the EC2 endpoint with port `5601`. By doing so, you can view and interact with all the saved objects in accordance with the FGAC settings and SAML authentication.
```

> 8. Post container restart you can access the self-managed dashboards by hitting the EC2 endpoint with port 5601. By doing so, you can view and interact with all the saved objects in accordance with the Fine-Grained Access Control settings and SAML authentication.

Reviewer 2: Please use a different/more precise word than "hitting".

Author: ok

> [!CAUTION]

> > If the endpoint is transitioned to self-managed dashboards and the customer intends to revert to the managed service dashboards endpoint, they must repeat the same procedure, which involves changing the kibana_url in the security configuration file back to the managed service dashboards endpoint. Until this change is made, the managed service dashboards endpoint will remain inaccessible.

Reviewer 2: ```suggestion
> If the endpoint is transitioned to self-managed Dashboards and the user intends to revert to the managed service Dashboards endpoint, they must repeat the same procedure, which involves changing the kibana_url in the security configuration file back to the managed service Dashboards endpoint. Until this change is made, the managed service Dashboards endpoint will remain inaccessible.
```

> [!NOTE]

> > When using docker in EC2 instance, the self-managed dashboards cannot be accessed over the internet. It is only accessible within the same VPC.

Reviewer 2: ```suggestion
> When using Docker in an EC2 instance, the self-managed Dashboards instance cannot be accessed over the internet. It is only accessible within the same VPC.
```

# Summary

> The self-managed dashboards workaround during upgrade minimizes dashboard downtime and impact on your business operations. The workaround also supports multiple authentication methods to support any specific authentication method required by the customers. You can find references about these subjects in the resources provided in the following section.

Reviewer 2: ```suggestion
The self-managed Dashboards workaround during upgrade minimizes downtime and impact to your business operations. The workaround also supports multiple authentication methods. You can find more information in the resources provided in the following section.
```
